src/middlewared/middlewared/plugins/virt/instances.py

```diff
@@ -0,0 +1,445 @@
+import aiohttp
+import errno
+from typing import Any, List
+
+from middlewared.service import (
+    CallError, CRUDService, ValidationErrors, filterable, job, private
+)
+from middlewared.utils import filter_list
+
+from middlewared.api import api_method
+from middlewared.api.current import (
+    VirtInstanceEntry,
+    VirtInstancesCreateArgs, VirtInstancesCreateResult,
+    VirtInstancesUpdateArgs, VirtInstancesUpdateResult,
+    VirtInstancesDeleteArgs, VirtInstancesDeleteResult,
+    VirtInstancesStateArgs, VirtInstancesStateResult,
+    VirtInstancesImageChoicesArgs, VirtInstancesImageChoicesResult,
+    VirtInstancesDeviceListArgs, VirtInstancesDeviceListResult,
+    VirtInstancesDeviceAddArgs, VirtInstancesDeviceAddResult,
+    VirtInstancesDeviceDeleteArgs, VirtInstancesDeviceDeleteResult,
+)
+from .utils import incus_call, incus_call_and_wait
+
+
+LC_IMAGES_SERVER = 'https://images.linuxcontainers.org'
+LC_IMAGES_JSON = f'{LC_IMAGES_SERVER}/streams/v1/images.json'
+
+
+class VirtInstancesService(CRUDService):
+
+    class Config:
+        namespace = 'virt.instances'
```

Comment:
@yocalebo minor thing, but should we rename to virt.instance, singular? It seems we prefer that as an unspoken rule for endpoints?